src/Functions/multiIf.cpp

```diff
@@ -239,6 +260,73 @@ class FunctionMultiIf final : public FunctionIfBase
 
         return res;
     }
+
+private:
+    static void executeShortCircuitArguments(ColumnsWithTypeAndName & arguments)
+    {
+        int last_short_circuit_argument_index = checkShirtCircuitArguments(arguments);
+        if (last_short_circuit_argument_index < 0)
+            return;
+
+        /// Let's denote x_i' = maskedExecute(x_i, mask).
+        /// multiIf(x_0, y_0, x_1, y_1, x_2, y_2, ..., x_{n-1}, y_{n-1}, y_n)
+        /// We will support mask_i = !x_0 & !x_1 & ... & !x_i
+        /// and condition_i = !x_0 & ... & !x_{i - 1} & x_i
+        /// Base:
+        /// mask_0 and condition_0 is 1 everywhere, x_0' = x_0.
+        /// Iteration:
+        /// condition_i = extractMask(mask_{i - 1}, x_{i - 1}')
+        /// y_i' = maskedExecute(y_i, condition)
+        /// mask_i = extractMask(mask_{i - 1}, !x_{i - 1}')
+        /// x_i' = maskedExecute(x_i, mask)
+        /// Also we will treat NULL as 0 if x_i' is Nullable.
+
+        IColumn::Filter mask(arguments[0].column->size(), 1);
+        MaskInfo mask_info = {.has_ones = true, .has_zeros = false};
+        IColumn::Filter condition_mask(arguments[0].column->size(), 1);
+        MaskInfo condition_mask_info = {.has_ones = true, .has_zeros = false};
+
+        int i = 1;
+        while (i <= last_short_circuit_argument_index)
+        {
+            auto & cond_column = arguments[i - 1].column;
+            /// If condition is const or null and value is false, we can skip execution of expression after this condition.
+            if ((isColumnConst(*cond_column) || cond_column->onlyNull()) && !cond_column->empty() && !cond_column->getBool(0))
+            {
+                condition_mask_info.has_ones = false;
+                condition_mask_info.has_zeros = true;
+            }
+            else
+            {
+                condition_mask_info = extractMask(mask, cond_column, condition_mask);
+                maskedExecute(arguments[i], condition_mask, condition_mask_info);
+            }
+
+            /// Check if the condition is always true and we don't need to execute the rest arguments.
+            if (!condition_mask_info.has_zeros)
+                break;
+
+            ++i;
+            if (i > last_short_circuit_argument_index)
+                break;
+
+            /// Extract mask only if it make sense.
+            if (condition_mask_info.has_ones)
+                mask_info = extractMask<true>(mask, cond_column, mask);
```

Comment:
Can we use `extractMaskInplace` here?

Reply:
Yes, we can, don't know why I didn't used it.